### PR TITLE
[opt](nereids)adjust distribution cost for better choice of broadcast join and shuffle join

### PR DESCRIPTION
**Pull request by englefly**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by englefly** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/statistics/Statistics.java [hunk 1]
```diff
@@ -129,7 +129,7 @@ public double computeSize() {
     }
 
     public double dataSizeFactor() {
-        return computeTupleSize() / K_BYTES;
+        return Math.max(computeTupleSize() / K_BYTES, 1);
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/Statistics.java:132]: please add explaination for this change and list the tpcds 1T perf data comparison, some cases can be put into consideration, such as q4/q11.

**Comment by englefly** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java [hunk 2]
```diff
@@ -457,6 +457,11 @@ public class SessionVariable implements Serializable, Writable {
             SHOW_HIDDEN_COLUMNS
     );
 
+    public static final String SHUFFLE_FACTOR_LOWER_BOUND = "shuffle_factor_lower_bound";
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:460]: remove this by using a fixed value (we have hint control as a no-brainer plan) or add upper side also

File: fe/fe-core/src/main/java/org/apache/doris/statistics/Statistics.java [hunk 3]
```diff
@@ -129,7 +130,12 @@ public double computeSize() {
     }
 
     public double dataSizeFactor() {
-        return computeTupleSize() / K_BYTES;
+        double lowerBound = 0.03;
+        double upperBound = 0.07;
+        if (ConnectContext.get() != null) {
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/Statistics.java:135]: leave this as fixed bound value or control them by session vars both

File: regression-test/data/nereids_ssb_shape_sf100_p0/shape/q3.4.out [hunk 4]
```diff
@@ -8,23 +8,22 @@ PhysicalResultSink
...
+------------------------------PhysicalOlapScan[supplier]
+----------------------PhysicalDistribute
+------------------------PhysicalProject
+--------------------------filter(c_city IN ('UNITED KI1', 'UNITED KI5'))
+----------------------------PhysicalOlapScan[customer]
 ------------------PhysicalDistribute
 --------------------PhysicalProject
-----------------------filter(c_city IN ('UNITED KI1', 'UNITED KI5'))
-------------------------PhysicalOlapScan[customer]
+----------------------filter((dates.d_yearmonth = 'Dec1997'))
```
Comment (xzj7019) [inline regression-test/data/nereids_ssb_shape_sf100_p0/shape/q3.4.out:27]: join order is changed, pls make sure the perf doesn't have regression

File: regression-test/data/nereids_tpcds_shape_sf1000_p0/shape/query10.out [hunk 5]
```diff
@@ -26,8 +26,7 @@ PhysicalResultSink
 ----------------------------hashJoin[LEFT_SEMI_JOIN] hashCondition=((c.c_customer_sk = web_sales.ws_bill_customer_sk)) otherCondition=()
 ------------------------------PhysicalProject
 --------------------------------hashJoin[INNER_JOIN] hashCondition=((customer_demographics.cd_demo_sk = c.c_current_cdemo_sk)) otherCondition=() build RFs:RF3 c_current_cdemo_sk->[cd_demo_sk]
-----------------------------------PhysicalDistribute
-------------------------------------PhysicalOlapScan[customer_demographics] apply RFs: RF3
+----------------------------------PhysicalOlapScan[customer_demographics] apply RFs: RF3
```
Comment (xzj7019) [inline regression-test/data/nereids_tpcds_shape_sf1000_p0/shape/query10.out:29]: Pls add perf diff detail of ds 100g/1t for a confirmation

File: regression-test/data/nereids_tpcds_shape_sf1000_p0/shape/query77.out [hunk 6]
```diff
@@ -93,19 +93,20 @@ PhysicalResultSink
...
---------------------------hashAgg[LOCAL]
-----------------------------PhysicalProject
-------------------------------hashJoin[INNER_JOIN] hashCondition=((web_returns.wr_web_page_sk = web_page.wp_web_page_sk)) otherCondition=() build RFs:RF7 wp_web_page_sk->[wr_web_page_sk]
---------------------------------PhysicalDistribute
-----------------------------------PhysicalProject
-------------------------------------hashJoin[INNER_JOIN] hashCondition=((web_returns.wr_returned_date_sk = date_dim.d_date_sk)) otherCondition=() build RFs:RF6 d_date_sk->[wr_returned_date_sk]
---------------------------------------PhysicalProject
-----------------------------------------PhysicalOlapScan[web_returns] apply RFs: RF6 RF7
---------------------------------------PhysicalDistribute
+--------------------------hashAgg[GLOBAL]
```
Comment (xzj7019) [inline regression-test/data/nereids_tpcds_shape_sf1000_p0/shape/query77.out:96]: Pls make sure the change from agg (1 stage changed to 2) is expected and perf has no regression.

File: regression-test/data/nereids_tpcds_shape_sf1000_p0/shape/query14.out [hunk 7]
```diff
@@ -4,11 +4,10 @@ PhysicalCteAnchor ( cteId=CTEId#0 )
 --PhysicalCteProducer ( cteId=CTEId#0 )
 ----PhysicalProject
 ------hashJoin[INNER_JOIN] hashCondition=((item.i_brand_id = t.brand_id) and (item.i_category_id = t.category_id) and (item.i_class_id = t.class_id)) otherCondition=() build RFs:RF6 class_id->[i_class_id];RF7 category_id->[i_category_id];RF8 brand_id->[i_brand_id]
+--------PhysicalProject
```
Comment (xzj7019) [inline regression-test/data/nereids_tpcds_shape_sf1000_p0/shape/query14.out:7]: Is the plan changed from bucket shuffle to broadcast? Pls make sure it is expected.

File: regression-test/data/nereids_tpcds_shape_sf100_p0/noStatsRfPrune/query35.out [hunk 8]
```diff
@@ -25,25 +25,23 @@ PhysicalResultSink
 --------------------------hashJoin[LEFT_SEMI_JOIN] hashCondition=((c.c_customer_sk = catalog_sales.cs_ship_customer_sk)) otherCondition=()
 ----------------------------hashJoin[LEFT_SEMI_JOIN] hashCondition=((c.c_customer_sk = web_sales.ws_bill_customer_sk)) otherCondition=()
 ------------------------------PhysicalProject
---------------------------------hashJoin[INNER_JOIN] hashCondition=((customer_demographics.cd_demo_sk = c.c_current_cdemo_sk)) otherCondition=()
+--------------------------------hashJoin[INNER_JOIN] hashCondition=((c.c_current_addr_sk = ca.ca_address_sk)) otherCondition=()
```
Comment (xzj7019) [inline regression-test/data/nereids_tpcds_shape_sf100_p0/noStatsRfPrune/query35.out:28]: join order is changed and pls make sure the perf has no regression

File: regression-test/data/nereids_tpcds_shape_sf100_p0/noStatsRfPrune/query38.out [hunk 9]
```diff
@@ -44,18 +44,16 @@ PhysicalResultSink
...
-----------------------------PhysicalProject
-------------------------------hashJoin[INNER_JOIN] hashCondition=((web_sales.ws_sold_date_sk = date_dim.d_date_sk)) otherCondition=() build RFs:RF4 d_date_sk->[ws_sold_date_sk]
---------------------------------PhysicalDistribute
-----------------------------------PhysicalProject
-------------------------------------PhysicalOlapScan[web_sales] apply RFs: RF4 RF5
---------------------------------PhysicalDistribute
-----------------------------------PhysicalProject
-------------------------------------filter((date_dim.d_month_seq <= 1194) and (date_dim.d_month_seq >= 1183))
---------------------------------------PhysicalOlapScan[date_dim]
+------------------------hashJoin[INNER_JOIN] hashCondition=((web_sales.ws_sold_date_sk = date_dim.d_date_sk)) otherCondition=() build RFs:RF5 d_date_sk->[ws_sold_date_sk]
```
Comment (xzj7019) [inline regression-test/data/nereids_tpcds_shape_sf100_p0/noStatsRfPrune/query38.out:47]: order is changed and pls make sure the perf has no regression

File: regression-test/data/nereids_tpcds_shape_sf1000_p0/shape/query64.out [hunk 10]
```diff
@@ -19,20 +19,18 @@ PhysicalCteAnchor ( cteId=CTEId#1 )
 --------------------------------PhysicalDistribute
 ----------------------------------PhysicalProject
 ------------------------------------hashJoin[INNER_JOIN] hashCondition=((customer.c_first_sales_date_sk = d2.d_date_sk)) otherCondition=() build RFs:RF15 d_date_sk->[c_first_sales_date_sk]
---------------------------------------PhysicalDistribute
-----------------------------------------PhysicalProject
-------------------------------------------hashJoin[INNER_JOIN] hashCondition=((customer.c_current_hdemo_sk = hd2.hd_demo_sk)) otherCondition=() build RFs:RF14 hd_demo_sk->[c_current_hdemo_sk]
+--------------------------------------PhysicalProject
```
Comment (xzj7019) [inline regression-test/data/nereids_tpcds_shape_sf1000_p0/shape/query64.out:22]: order is changed.

**Comment by englefly** [comment]
run buildall

File: regression-test/data/nereids_ssb_shape_sf100_p0/shape/q3.4.out (hunk 4 above)
Comment (englefly) [reply regression-test/data/nereids_ssb_shape_sf100_p0/shape/q3.4.out:27]: improved from 0.13 sec to 0.11 sec.
this join order change has no impact on performance. the improvement comes from the broadcast join.

**Review by xzj7019 — APPROVED** [review]

**Review by starocean999 — APPROVED** [review]

**Pull request merged**